src/compiler/checker.ts

```diff
@@ -28753,7 +28753,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (!hasDefaultClause) {
                 return caseType;
             }
-            const defaultType = filterType(type, t => !(isUnitLikeType(t) && contains(switchTypes, getRegularTypeOfLiteralType(extractUnitType(t)))));
+            const defaultType = filterType(type, t => !(isUnitLikeType(t) && contains(switchTypes, t.flags & TypeFlags.Undefined ? undefinedType : getRegularTypeOfLiteralType(extractUnitType(t)))));

```

Comment:
a potential alternative to this could look like this:
```suggestion
            const switchTypesUnion = getUnionType(filter(switchTypes, isUnitType));
            const defaultType = filterType(type, t => !(isUnitLikeType(t) && areTypesComparable(t, switchTypesUnion)));
```

this is inspired by the logic used by `narrowTypeByEquality` and I've checked that this alternative would still pass the existing test cases.